### ISSUE TITLE
Screen reader accessibility for the supercollider IDE

### ISSUE THREAD

**Issue by readmemyrights** [supercollider/supercollider#7541]
# Motivation

I have suggested an accessibility related improvement in #7527, for an option to disable the currently inaccessible and thus more distracting autocompletion popups. Since there fortunately seems to be interest in improving the accessibility of the IDE I decided to make an issue that puts them together in one place.

# Description

I have sorted them by importance, the more important ones, at least in my opinion, coming first, and the less important ones or the ones for which work arounds exist coming towards the end.

The accessibility of a program varies a lot by the operating system, screen reader, version of the program and the screen reader. I am at the moment able to test with NVDA on windows and cthulhu and orca screen readers on linux. Testing from people who use different combinations of OS and screen reader would be appreciated.

# The issues

## Help system

The most critical one by far. On windows it was possible until a certain version to at least tab through the links and edit fields and use certain commands to help read the text, but now you can barely get the focus on the help screen to do ctrl+a and ctrl+c to paste it into notepad, the only workaround besides searching for an online reference. The situation is similar on linux.

I heard there was a rework on the help system in progress, but I haven't gotten a chance to read what exactly it implies. In general QT's webview widgets leave a lot to be desired as far as accessibility.

A potential solution is to open an external web browser, which one can rely on to be accessible, to display the help page instead of the internal help view.

Minor issue with the html help: the edit boxes with the code examples are for some reason shown as one line with some strange symbol where the new lines should be, the screen reader can't actually interact with the edit box I can only copy and paste it into notepad to read it normally. Is there something special these edit boxes do besides displaying text? If so it could likely be replaced by a [<textarea> element](https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Elements/textarea).

Where can I download older supercollider versions? I would like to test which version introduced the regression in the help system and what it could be related to.

# Quarks screen

The list is inaccessible, it only reads out the button to select / deselect it. I tested this on windows, will eventually try it on linux as well.

## Edit field

On windows, blank lines are read the same as the closest nonempty line above it. This issue isn't unique to supercollider unfortunately, it seems to be common in QT applications.

Linux doesn't have this problem, but it has a different one where pressing backspace reads every deleted character like 'space" instead of the deleted character. I'll test it later with other applications to see if it's another QT issue.

## Post window

On windows, the post window can be accessed, but arrow keys don't work inside it, instead one needs to reads it's contents with the review cursor. Supercollider isn't the only one that has this issue, mudlet and Sonic PI are two other applications that have a similar problem. I assume it's on QT's end.

On linux, post window acts similarly, except that using it's rough equivalent of review mode doesn't work. However, the screen reader reads the contents of the post window automatically, which works for most cases. I will look into the review issue more, since it could be a problem with me not understanding the screen readers on linux as well.

## General recommendations

I'm not a qt programmer, so keep in mind that nothing in this section is definitive especially if it talks about qt widgets. I'm just guessing the situation is similar to web development, where similar problems often come up.

In most cases, accessibility issues come about because of widgets that are created specifically for a program instead of using one of the standard ones. GUI toolkits are written so that the most common widgets they offer work accessibly out of the box, and because of how common they are screen reader users know what to expect from them.

However, it's also likely that QT is responsible for some of the mistakes, especially when the issue seems to be common between many different programs. Patching it with a new widget could work, but it could also introduce new issues along the way.

**Comment by JordanHendersonMusic** [comment]
@readmemyrights thanks for this! I'm not aware of any other regular developers who use screen readers. Would you mind just explaining how they work a little? What information does it consume? How do you navigate? And what software would you recommend? - or if you have a link to a good explanation?

If it turns out we can generate what ever information the screen reader consumes, I'm wondering if we might be able to track how this changes with each commit and prevent any further regressions?

**Comment by readmemyrights** [comment]
Hello,

Thanks for the quick response, and sorry for not responding before.

I don't know the specifics, but there are APIs by which interface libraries like QT, GTK, etc give information to screen readers, which then read it out using a speech synthesizer and let the user navigate and get more information using keyboard shortcuts. On windows the api seems to be [Active accessibility](https://learn.microsoft.com/en-us/windows/win32/winauto/microsoft-active-accessibility), on linux it's [ATSPI2](https://docs.gtk.org/atspi2/) (it's a GTK link but applies to other libraries too), and the equivalent on MacOS escapes me at the moment, but it all works by the same principal: there's a way for graphical programs to expose widgets on the screen, and screen readers get information about it.

I am not a GUI developer like I said, but I'll try to ask around to see if anyone else tried something like what you're suggestion. main issue I see is that it's too low level, and it's something that QT already handles pretty well for the most part. Unless supercollider makes it's own brand new widgets it likely doesn't need to bother with these APIs directly. In case it does, I found [this page in QT's documentation](https://doc.qt.io/qt-6/accessible-qwidget.html) that should hopefully point in the right direction.

Sorry I couldn't give a more prompt or direct response. Like I said in the original post, best bet is to either stick to standard tested widgets as much as possible, or give options to make parts of it work differently, like the option to open the local help in a web browser I suggested.
